File: android/src/main/java/org/ligi/gobandroid_hd/ui/application/navigation/NavigationDrawerHandler.java
package org.ligi.gobandroid_hd.ui.application.navigation;

import android.content.Intent;
import android.net.Uri;
import android.support.design.widget.NavigationView;
import android.view.MenuItem;

import org.greenrobot.eventbus.EventBus;
import org.ligi.axt.AXT;
import org.ligi.gobandroid_hd.App;
import org.ligi.gobandroid_hd.R;
import org.ligi.gobandroid_hd.events.GameChangedEvent;
import org.ligi.gobandroid_hd.logic.GoGame;
import org.ligi.gobandroid_hd.model.GameProvider;
import org.ligi.gobandroid_hd.ui.BaseProfileActivity;
import org.ligi.gobandroid_hd.ui.GoPrefsActivity;
import org.ligi.gobandroid_hd.ui.UnzipSGFsDialog;
import org.ligi.gobandroid_hd.ui.application.GobandroidFragmentActivity;
import org.ligi.gobandroid_hd.ui.application.GobandroidSettings;
import org.ligi.gobandroid_hd.ui.links.LinksActivity;
import org.ligi.gobandroid_hd.ui.recording.GameRecordActivity;
import org.ligi.gobandroid_hd.ui.sgf_listing.SGFFileSystemListActivity;

import java.io.File;

import javax.inject.Inject;

public class NavigationDrawerHandler {

    private final GobandroidFragmentActivity ctx;
    private final NavigationView navigationDrawer;

    @Inject
    GobandroidSettings settings;

    @Inject
    GameProvider gameProvider;

    public NavigationDrawerHandler(final GobandroidFragmentActivity ctx) {
        App.component().inject(this);
        this.ctx = ctx;

        navigationDrawer = (NavigationView) ctx.findViewById(R.id.left_drawer);
    }

    public void handle() {
        navigationDrawer.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(MenuItem item) {
                switch (item.getItemId()) {

                    case R.id.menu_drawer_empty:
                        final GoGame act_game = gameProvider.get();

                        gameProvider.set(new GoGame((byte) act_game.getSize(), (byte) act_game.getHandicap()));

                        EventBus.getDefault().post(GameChangedEvent.INSTANCE);

                        ctx.startActivity(new Intent(ctx, GameRecordActivity.class).setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP));
                        return true;


                    case R.id.menu_drawer_links:
                        ctx.startActivity(new Intent(ctx, LinksActivity.class).setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP));
                        return true;

                    case R.id.menu_drawer_settings:

                        ctx.startActivity(new Intent(ctx, GoPrefsActivity.class).setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP));
                        return true;


                    /*case R.id.online_play:
                        CloudHooks.profileOrOnlinePlay(ctx);
                        return true;
*/

                    case R.id.menu_drawer_tsumego:
                        Intent next = startSGFListForPath(settings.getTsumegoPath());

                        if (!unzipSGFifNeeded(next)) {
                            ctx.startActivity(next);
                        }

                        return true;

                    case R.id.menu_drawer_review:
                        Intent next2 = startSGFListForPath(settings.getReviewPath());

                        if (!unzipSGFifNeeded(next2)) {
                            ctx.startActivity(next2);
                        }

                        return true;

                    case R.id.menu_drawer_bookmark:
                        ctx.startActivity(startSGFListForPath(settings.getBookmarkPath()));

                        return true;


                    case R.id.menu_drawer_profile:
                        ctx.startActivity(new Intent(ctx, BaseProfileActivity.class));

                        return true;


                    case R.id.menu_drawer_beta:
                        AXT.at(ctx).startCommonIntent().openUrl("https://play.google.com/apps/testing/org.ligi.gobandroid_hd");
                        return true;

                }
                return false;
            }
        });
    }

    private Intent startSGFListForPath(File path) {
        final Intent i = new Intent(ctx, SGFFileSystemListActivity.class);
        i.setData(Uri.parse("file://" + path.getAbsolutePath()));
        return i;
    }

    /**
     * Downloads SGFs and shows a ProgressDialog when needed
     *
     * @return - weather we had to unzip files
     */
    public boolean unzipSGFifNeeded(Intent intent_after) {
        // we check for the tsumego path as the base path could already be there but  no valid tsumego

        final File tsumegoPath = settings.getTsumegoPath();
        if (!tsumegoPath.isDirectory() || settings.isVersionSeen(1)) {
            new UnzipSGFsDialog(ctx, intent_after, settings).show();
            return true;
        }
        return false;

    }

}

File: android/src/main/java/org/ligi/gobandroid_hd/ui/application/navigation/NavigationDrawerHandler.kt
package org.ligi.gobandroid_hd.ui.application.navigation

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.support.design.widget.NavigationView
import com.github.salomonbrys.kodein.instance
import com.github.salomonbrys.kodein.lazy
import kotlinx.android.synthetic.main.navigation_drawer_container.*
import org.greenrobot.eventbus.EventBus
import org.ligi.gobandroid_hd.App
import org.ligi.gobandroid_hd.R
import org.ligi.gobandroid_hd.events.GameChangedEvent
import org.ligi.gobandroid_hd.logic.GoGame
import org.ligi.gobandroid_hd.model.GameProvider
import org.ligi.gobandroid_hd.ui.BaseProfileActivity
import org.ligi.gobandroid_hd.ui.GoPrefs
import org.ligi.gobandroid_hd.ui.GoPrefsActivity
import org.ligi.gobandroid_hd.ui.UnzipSGFsDialog
import org.ligi.gobandroid_hd.ui.application.GoAndroidEnvironment
import org.ligi.gobandroid_hd.ui.application.GobandroidFragmentActivity
import org.ligi.gobandroid_hd.ui.links.LinksActivity
import org.ligi.gobandroid_hd.ui.recording.GameRecordActivity
import org.ligi.gobandroid_hd.ui.sgf_listing.SGFFileSystemListActivity
import org.ligi.kaxt.startActivityFromURL
import java.io.File

class NavigationDrawerHandler(private val ctx: GobandroidFragmentActivity) {

    internal val env: GoAndroidEnvironment by App.kodein.lazy.instance()
    internal val gameProvider: GameProvider by App.kodein.lazy.instance()


    val actionMap by lazy {
        mapOf(
                R.id.menu_drawer_empty to {
                    val act_game = gameProvider.get()
                    gameProvider.set(GoGame(act_game.size.toByte().toInt(), act_game.handicap.toByte().toInt()))
                    EventBus.getDefault().post(GameChangedEvent)
                    startForClass(GameRecordActivity::class.java)
                },

                R.id.menu_drawer_links to {
                    startForClass(LinksActivity::class.java)
                },

                R.id.menu_drawer_settings to {
                    startForClass(GoPrefsActivity::class.java)
                },

                R.id.menu_drawer_tsumego to {
                    startForPath(env.tsumegoPath)
                },

                R.id.menu_drawer_review to {
                    val path = env.reviewPath
                    startForPath(path)

                },

                R.id.menu_drawer_bookmark to {
                    ctx.startActivity(startSGFListForPath(env.bookmarkPath))
                },


                R.id.menu_drawer_profile to {
                    ctx.startActivity(Intent(ctx, BaseProfileActivity::class.java))
                },

                R.id.menu_drawer_beta to {
                    ctx.startActivityFromURL("https://play.google.com/apps/testing/org.ligi.gobandroid_hd")
                },

                R.id.menu_drawer_translation to {
                    ctx.startActivityFromURL("https://www.transifex.com/ligi/gobandroid")
                }
        )
    }

    private fun startForClass(java: Class<out Activity>) {
        ctx.startActivity(Intent(ctx, java).setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP))
    }

    private fun startForPath(path: File) {
        val next = startSGFListForPath(path)

        if (!unzipSGFifNeeded(next)) {
            ctx.startActivity(next)
        }
    }

    fun handle() {
        ctx.left_drawer.setNavigationItemSelectedListener(NavigationView.OnNavigationItemSelectedListener { item ->
            val function = actionMap[item.itemId]
            if (function != null) {
                ctx.closeDrawers()
                function.invoke()
                return@OnNavigationItemSelectedListener true
            }
            return@OnNavigationItemSelectedListener false
        })
    }


    private fun startSGFListForPath(path: File) = Intent(ctx, SGFFileSystemListActivity::class.java).apply {
        data = Uri.parse("file://" + path.absolutePath)
    }

    /**
     * Downloads SGFs and shows a ProgressDialog when needed

     * @return - weather we had to unzip files
     */
    fun unzipSGFifNeeded(intent_after: Intent): Boolean {
        // we check for the tsumego path as the base path could already be there but  no valid tsumego

        val tsumegoPath = env.tsumegoPath
        if (!tsumegoPath.isDirectory || GoPrefs.isVersionSeen(2)) {
            UnzipSGFsDialog(ctx, intent_after, env).show()
            return true
        }
        return false

    }

}
